### PR TITLE
Use window resurrection

### PR DESCRIPTION
**Pull request by AndyDyer**
resurrecting #80  since the original author last had activity on it in Jan 2021

### CODE REVIEW

**Comment by Andrey-Bazhanov** [comment]
Hi @tannerlinsley, please take a look on this PR.  There are lot peoples waiting for this feature.

**Review by centrodph — APPROVED** [review]

**Comment by Zertz** [comment]
For what it's worth, I switched our `react-table` implementation from `react-window` + `react-window-scroller` to this branch and it  works very well! I was even able to remove a couple hacks that were necessary before. 👌

**Comment by Zertz** [comment]
I needed build artifacts to use this so I forked and published this PR on npm under [react-virtual-window-fork](https://www.npmjs.com/package/react-virtual-window-fork), with hopefully clear attribution to @tannerlinsley, @AndyDyer and @petersendidit

Hopefully this gets merged eventually 🙏

**Review by piecyk — COMMENTED** [review]
Overall looks good. Wondering now, as this is a low level hook, do we want to include it? On other hand useVirtualWindow is pretty common case 🤔

**Comment by tannerlinsley** [comment]
This PR is up next. Looking through it now.

File: src/useWindowRect.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import React from 'react'
+
+import useIsomorphicLayoutEffect from './useIsomorphicLayoutEffect'
+
+export default function useWindowRect(windowRef) {
```
[the remark this answers is not in the document]
Comment (tannerlinsley) [reply src/useWindowRect.js:5]: At the very least, let's make this change. I'm concerned with the added bundle size for this addition, so it's a good start.

**Comment by tannerlinsley** [comment]
I'd love to get everyone to put their 👀 on this PR one last time and see if we could possibly rethink the way we support the window element. It *seems* like we should have almost all of the utilities and abstractions we need to make this happen with fewer changes. Any ideas?

- Could we simply pass different hooks and elements to `useVirtual`?

**Review by tannerlinsley — CHANGES_REQUESTED** [review]

**Comment by piecyk** [comment]
> Could we simply pass different hooks and elements to useVirtual?

I think we could, maybe like this https://github.com/tannerlinsley/react-virtual/pull/219

**Comment by AndyDyer** [comment]
> > Could we simply pass different hooks and elements to useVirtual?
> 
> I think we could, maybe like this #219

I like this

**Comment by Zertz** [comment]
For what it's worth, it would be extremely useful to have a single hook that can handle both use cases. We have a virtual table component that, for various reasons, must include both element-relative and window-relative "modes" and having separate hooks means we set one to the actual data size and the other to zero. This works but it would be neat if we could switch between these modes just by conditionally swapping the reference element.

Either way, great work everyone and I'll happily take whichever API you pick! 🎉

**Comment by piecyk** [comment]
> For what it's worth, it would be extremely useful to have a single hook that can handle both use cases.

Having `useScroll` abstraction, it would be possible to write that, for example one option could look like, imho much more flexible

```jsx
 const rowVirtualizer = useVirtual({
  size: rows.length,
  parentRef,
  useScroll: props => {
    const elementRes = useElementScroll(props)

    const windowRes = useWindowScroll({ ...props, windowRef: React.useRef(window) }),

    return isWindow ? windowRes : elementRes
  }
})
```

**Comment by piecyk** [comment]
To directly support window scroll, we need to have some kind of boolean in which mode is the useVirtual, for example isWindow, and the ref to window, windowRef. 

```tsx
const rowVirtualizer = useVirtual({
  size: rows.length,
  parentRef,
  isWindow,
  windowRef: React.useRef(window),
})
  ```
  
fyi Update the #219 to include those changes, build on top of useScroll

**Comment by Zertz** [comment]
Would it be possible for `parentRef` and `windowRef` to be mutually exclusive and then use that internally to determine the mode we're in?

**Comment by piecyk** [comment]
> Would it be possible for parentRef and windowRef to be mutually exclusive and then use that internally to determine the mode we're in?

Yes, as windowRef is optional, we can check if `windowRef !== undefined` to determine if in window scroll mode.

**Comment by circlingthesun** [comment]
> > Would it be possible for parentRef and windowRef to be mutually exclusive and then use that internally to determine the mode we're in?
> 
> Yes, as windowRef is optional, we can check if `windowRef !== undefined` to determine if in window scroll mode.

Surely you can just pass a window ref as the parentRef and do a check inside to hook to determine if its a window ref ( `parentRef.current === window` )? That'll save you two parameters. I did exactly that here: https://github.com/circlingthesun/react-virtual/blob/master/index.tsx#L44

**Comment by piecyk** [comment]
@circlingthesun what about scrollToIndex? Using `parentRef` we can calculate the offset of the list and correctly scroll to index. Imho having explicit windowRef is better in long run.

**Comment by circlingthesun** [comment]
@piecyk useRect was also updated see: https://github.com/circlingthesun/react-virtual/blob/master/use-rect.tsx#L6.

**Comment by circlingthesun** [comment]
I've been running this in production for over a year and scrollToIndex works, at least for my purposes. I was going to make a PR but my fork was based on a previous version and, at the time, it seemed like the previous PR was about to be accepted. I prefer a less complex interface. I'm just raising this to point out another approach.

**Comment by piecyk** [comment]
Guys the #219 is ready for review, as mention before imho `useScroll` is very flexible, and allow to change the default implementation very easy. Tested the window scroll with dynamic list/grid and works pretty good.

**Comment by AndyDyer** [comment]
> Surely you can just pass a window ref as the parentRef and do a check inside to hook to determine if its a window ref ( `parentRef.current === window` )? That'll save you two parameters. I did exactly that here: https://github.com/circlingthesun/react-virtual/blob/master/index.tsx#L44

This makes total complete sense and definitely slims down the API

**Comment by tannerlinsley** [comment]
This looks fantastic. Any way we could get a simple test set up for this? Even if it doesn't go super deep, I just want to make sure things at least render with the new hook.

**Comment by piecyk** [comment]
@AndyDyer we need the parentRef to recognise where it's in page, and take it under account when scrolling.

@tannerlinsley added some simple tests 🙈

**Comment by tannerlinsley** [comment]
Awesome. I merged the other PR, so this should be ready to fix up

**Comment by tannerlinsley** [comment]
I think we might need to point this PR at the `next` branch or just open a new PR. If no one responds here otherwise, I'll close this PR in a few days.

**Comment by lironezra** [comment]
Hi Guys,

Anyone knows what going on with this issue?
There is another way to achieve the virtual scrolling with `window` ref?

Thanks!!!

**Comment by tannerlinsley** [comment]
This PR is outdated and should either be closed, re-architected or rebased against v3 beta. Thanks!

**Comment by epurban** [comment]
Hey @tannerlinsley, is this window scrolling functionality attainable currently?

**Comment by piecyk** [comment]
> Hey @tannerlinsley, is this window scrolling functionality attainable currently?

@epurban yes it is in latest beta, you can use the `useWindowVirtualizer`

**Pull request closed without merging**